### PR TITLE
[fix] prevent context leak

### PR DESCRIPTION
**Pull request by asavah**
## Description

Fix context leak introduced in #16314 

It looks like that a context opened by avcodec_input_open() can't be reused,
otherwise it causes a leak.

## Motivation and Context

See discussion in https://github.com/xbmc/xbmc/pull/16314#issuecomment-514271843

## How Has This Been Tested?

Has been in my private builds of master for the last 2 weeks.

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp
Comment (Rechi) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: one space before `=` is enough

**Comment by asavah** [comment]
@Rechi 
This was copied verbatim from https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp#L236
Do you want me to fix whitespace only for the added code or should I fix whitespace in the original entry too?
Do you need a backport to Leia?

**Comment by phunkyfish** [comment]
Just fix the white space in the code you added.

On the backport the Leia PR for the streaminfo changes is not in Leia yet so we can incorporate your changes there once merged to master.

**Comment by asavah** [comment]
Done

**Comment by phunkyfish** [comment]
@FernetMenta are you happy to approve the PR?

**Comment by FernetMenta** [comment]
LGTM

**Comment by phunkyfish** [comment]
@hudokkow I think this is good to go in once someone approves it.

@asavah is it ok with you if I just add the new lines to the existing backport PR?

**Review by ksooo — APPROVED** [review]
Approving on behalf of @FernetMenta

**Comment by asavah** [comment]
@phunkyfish 
> is it ok with you if I just add the new lines to the existing backport PR?

Ofcourse it is ), thanks.

**Comment by phunkyfish** [comment]
> > is it ok with you if I just add the new lines to the existing backport PR?
> 
> Ofcourse it is ), thanks.

Done, backport combined with parent backport PR: #16355

**Comment by MilhouseVH** [comment]
Hi, as reported on the [forum](https://forum.kodi.tv/showthread.php?tid=343069&pid=2873940#pid2873940) this change is causing a crash when playing certain video streams, eg. `rtsp://184.72.239.149/vod/mp4:BigBuckBunny_115k.mov`.

crash log: https://paste.kodi.tv/erenehemow.kodi

Reverting this PR stops kodi crashing and the video plays normally.

Reproducible with LE RPi, LE x86_64 and a plain Ubuntu Kodi build (latest kodi master).

**Comment by asavah** [comment]
@MilhouseVH This PR fixes context leak issue, it's well above my knowledge  to fix the reported crash, I'll take a look tho.
The only fix I currently see is to revert both this PR and #16314 .

**Comment by phunkyfish** [comment]
So the above video works with #16314 and without this PR?

**Comment by FernetMenta** [comment]
ffmpeg issue in rtsp_send_cmd_with_content_async. you could do some tweaks in kodi to work around this issue baut that's obviously nonsense.

**Comment by phunkyfish** [comment]
What’s the right way to fix it?

**Comment by MilhouseVH** [comment]
> So the above video works with #16314 and without this PR?

Yes, that is correct - latest Kodi master 6aba73be78e8228195c3326590d191a2c65760c4 with PR16395 reverted will play the test stream without crashing.

**Comment by asavah** [comment]
I think that in case of rtsp we should rise a flag, wait for interrupt_cb to fire, if flag is set - return 1 in interrupt_cb and after that it should be safe to call avformat_close_input(), but I fear I won't be able to implement this properly ...

**Comment by FernetMenta** [comment]
> What’s the right way to fix it?

https://github.com/xbmc/FFmpeg/blob/master/libavformat/rtspdec.c#L61

don't send teardown if rtsp_hd_out is null

**Comment by phunkyfish** [comment]
Thanks, will test and PR it later.

**Comment by asavah** [comment]
Thanks.

**Comment by FernetMenta** [comment]
> Thanks, will test and PR it later.

Note that the folks at ffmpeg still use an antiquated mailing list for submitting patches.

**Pull request merged**
